Filter "real" attachments by being referenced

pabzm opened the pull request:
This changes the way in which attachments are determined to be shown as downloadable ("standalone"), or not ("inline").
In theory this should be determined by their `Content-Disposition`, but in reality this often doesn't work.

Previously, on top of the default standard handling, edge cases have been fixed by checking for their concrete details.

Now the code checks if the `Content-ID` or `Content-Location` of the attachment is actually being used in other parts of the message. If not, the attachment is considered to be "standalone". This might make some images be shown as downloadable, which have a "inline"-disposition – but it'll happen only in cases where the message is broken, or they wouldn't have been shown at all previously. So I consider this a progression while staying a comprehensible and review-able changeset. 

This fixes #9443. 

I didn't find an email that it breaks. I also don't see how it could.

It helps #9465, too.

And it (unintendedly) also helps #9326, because now the information, which image is actually referenced in an HTML-part, ist available already after message parsing.

pabzm committed:
Filter "real" attachments by being referenced

This changes the way in which attachments are determined to be shown as
such ("standalone"), or not ("inline").
In theory this should be determined by their Content-Disposition, but in
reality this often doesn't work.
Now we check if the Content-ID or Content-Location of the attachment is
actually being used in other parts of the message. If not, the
attachment is considered to be "standalone".

alecpl commented on program/lib/Roundcube/rcube_message.php:
```diff
@@ -569,6 +577,44 @@ public function is_attachment($part)
+     *
+     * @return bool True if the part is an attachment part
+     */
+    public function is_standalone_attachment(rcube_message_part $part): bool
```
Word "standalone" is not needed here. Also what "(or part of it)" means?

pabzm replied:
I didn't understand that description anymore, either, so I fixed that. 🤔

I chose "standalone" to stress the meaning, because asking if an attachment "is_attachment()" feels confusing to me.

If we use this code to distinctly decide whether a mime-part is an attachment *(x)or* an inline-part (see #9473), then I'd agree that we don't use the extra word (and probably don't even need this method anymore). How about that?

alecpl replied:
I just noticed that we already have `is_attachments()` method. Anyway, maybe the method name should be `is_referred_attachment()`.

pabzm replied:
I like that name, thank you for the suggestion! It allows to get rid of the negation in the calling code, too.

alecpl commented on program/lib/Roundcube/rcube_message.php:
```diff
@@ -569,6 +577,44 @@ public function is_attachment($part)
+    public function is_standalone_attachment(rcube_message_part $part): bool
+    {
+        $references = $this->get_replacement_references();
+        // This code is intentionally verbose to keep it comprehensible.
```
I prefer an empty line in here.

alecpl commented on program/lib/Roundcube/rcube_message.php:
```diff
@@ -569,6 +577,44 @@ public function is_attachment($part)
+        if (!empty($part->content_id) && in_array($part->content_id, $references)) {
+            return false;
+        }
+        // Filter out attachments that are reference by their
```
I prefer an empty line in here.

alecpl commented on program/lib/Roundcube/rcube_message.php:
```diff
@@ -569,6 +577,44 @@ public function is_attachment($part)
         return false;
     }
 
+    public function get_replacement_references(): array
```
Does it needs to be public? Also, the name is not great, but I have no better idea now. Finally, this should probably take `context` property into account. I.e. testing this with forwarded messages might be good idea.

pabzm replied:
Technically this could be `protected`, but it could be useful to solve #9326 as long as it's public.

pabzm replied:
I realized that to solve #9326 one would want to call `is_referred_attachment()`, too. So I changed the method's visibility.

alecpl commented:
Looks to me that it fixes #9443 only partially. With `inline_images=true` I expect the image listed on attachments list to be displayed not only on the list, but also below the body (here a non-existent body).

I think for `inline_images` feature we'll really need to look into the body and display the image when: 1) there's no text parts at all, 2) there's no HTML part, 3) there's HTML part and it does not have a reference to the image part.

pabzm committed:
Consider all mime-parts to check if message is empty

Previously only `parts` and `body` were checked, so mime-parts that were
classified into `attachments` and `inline_parts` didn't count – thus
messages that contained only those parts were shown blank.

pabzm commented:
> Looks to me that it fixes #9443 only partially

That's because the message was considered empty and thus no body was rendered. An issue that is not exactly in scope of this PR, but because it is related to handling mime-parts I fixed it nonetheless in here.

Now the image is shown inline as expected.

I'd be happy to clean up the commit history before merging if you'd agree.

pabzm commented:
Wait a moment, I'll have to rebase to "master" anyway, and then I can clean the history a little, too.

pabzm commented:
@alecpl I rebased onto the latest master. Could you merge this or let me know what's blocking it, please?

pabzm commented:
@alecpl Just a note: I would prefer if you wouldn't use squash-commits when merging, because it makes rebasing much harder (git can't automatically identify the commits that were used before). If you don't like a list of commits please just ask to clean it up.
